### DISCUSSION TITLE
Parsing/converting strings to enums

### DISCUSSION THREAD

**Discussion by CNBoland** [AutoMapper/AutoMapper.Extensions.EnumMapping#23] [Q&A]
I have a need to parse/convert strings to enums during integration between systems. The downstream system gives us strings that are converted to enums for processing. Using the example from documentation, the conversion sequence looks like this:

`"S"` => `Source.Second` => `Destination.Second`

I understand how to map `Source.Second` => `Destination.Second`, but I'm stuck on `"S"` => `Source.Second`.

I've tried `.CreateMap<string, Source>().ConvertUsingEnumMapping(opt => opt.MapByName("S", Source.Second))`, but the see this error:

`The type 'string' must be a non-nullable value type in order to use it as parameter 'TSource' in the generic type or method 'EnumMappingExpressionExtensions.ConvertUsingEnumMapping<TSource, TDestination>(IMappingExpression<TSource, TDestination>, Action<IEnumConfigurationExpression<TSource, TDestination>>)'`

Is support for this scenario built in to `EnumMapping`, or is something like [Custom Type Converter](https://docs.automapper.org/en/latest/Custom-type-converters.html) a better solution?

**Comment by HenkKin** [comment]
Maybe this example will help you?

https://dotnetfiddle.net/NGEXhs

```
using System;
using AutoMapper;
using AutoMapper.Extensions.EnumMapping;
					
public class Program
{
	public static void Main()
	{
		var configuration = new MapperConfiguration(cfg => 
		{
			cfg.CreateMap<string, SrcEnum>().ConvertUsing((string src, SrcEnum dest) => {
				// Just to show it works
				if(src=="C")
				{
					src= "D";
				}
				return Enum.Parse<SrcEnum>(src);
			});
			
			cfg.CreateMap<SrcEnum, DestEnum>().ConvertUsingEnumMapping(opt => opt
				.MapByName()
				.MapValue(SrcEnum.A, DestEnum.E)
				.MapValue(SrcEnum.B, DestEnum.F)
				.MapValue(SrcEnum.C, DestEnum.G)
				.MapValue(SrcEnum.D, DestEnum.H));
		});
		// only during development, validate your mappings; remove it before release
		configuration.AssertConfigurationIsValid();
		// use DI (http://docs.automapper.org/en/latest/Dependency-injection.html) or create the mapper yourself
		var mapper = configuration.CreateMapper();
		
		var input = "C";
		var result1 = mapper.Map<SrcEnum>(input);	
		var result2 = mapper.Map<DestEnum>(result1);	
		Console.WriteLine($"Input value: {input.GetType().Name} \"{input}\"");
		Console.WriteLine($"Mapped value 1: {result1.GetType().Name}.{result1}");
		Console.WriteLine($"Mapped value 2: {result2.GetType().Name}.{result2}");
	}		
}

public enum SrcEnum
{
	A, B, C, D
}
public enum DestEnum
{
	E, F, G, H
}

```

The output will be:
```
Input value: String "C"
Mapped value 1: SrcEnum.D
Mapped value 2: DestEnum.H
```

**Reply by CNBoland** [reply]
@HenkKin That works nicely. Thank you!
